concourse/steps/release.py

```diff
@@ -49,166 +350,98 @@ def release_and_prepare_next_dev_cycle(
-    # Bump release version and add suffix
-    next_version = version.process_version(
-        version_str=release_version,
-        operation=version_operation
-    )
-    next_version_dev = version.process_version(
-        version_str=next_version,
-        operation='set_prerelease',
-        prerelease=prerelease_suffix
+    release_tag_step = ReleaseTagStep(
+        github_helper=github_helper,
+        release_version=release_version,
+        author_email=author_email,
+        author_name=author_name,
+        release_context=release_context,
     )
 
-    release_commit_sha = _create_and_push_release_commit(
-        repo_dir=repo_dir,
+    github_release_step = GitHubReleaseStep(
+        github_helper=github_helper,
         git_helper=git_helper,
-        version_file_path=repository_version_file_path,
+        githubrepobranch=githubrepobranch,
         release_version=release_version,
-        target_ref=repository_branch,
-        commit_msg=f'Release {release_version}',
-        release_commit_callback=release_commit_callback,
+        component_descriptor_file_path=component_descriptor_file_path,
+        release_context=release_context,
     )
 
-    helper.create_tag(
-        tag_name=release_version,
-        tag_message="Release " + release_version,
-        repository_reference=release_commit_sha,
-        author_name=author_name,
-        author_email=author_email
-    )
-    release = helper.create_release(
-        tag_name=release_version,
-        body=release_notes_md,
-        draft=False,
-        prerelease=False
+    dev_cycle_commit_step = DevCyclePrepareStep(
+        github_helper=github_helper,
+        repo_dir=repo_dir,
+        repository_version_file_path=repository_version_file_path,
+        release_version=release_version,
+        version_operation=version_operation,
+        prerelease_suffix=prerelease_suffix,
+        release_context=release_context,
     )
 
-    if component_descriptor_file_path:
-        with open(component_descriptor_file_path) as f:
-            # todo: validate descriptor
-            component_descriptor_contents = f.read()
-        release.upload_asset(
-            content_type='application/x-yaml',
-            name=product.model.COMPONENT_DESCRIPTOR_ASSET_NAME,
-            asset=component_descriptor_contents,
-            label=product.model.COMPONENT_DESCRIPTOR_ASSET_NAME,
-        )
+    steps = [
+        release_commit_step,
+        release_tag_step,
+        github_release_step,
+        dev_cycle_commit_step,
+    ]
 
-    # Prepare version file for next dev cycle
-    helper.create_or_update_file(
-        file_path=repository_version_file_path,
-        file_contents=next_version_dev,
-        commit_message="Prepare next dev cycle " + next_version_dev
-    )
+    # validate all steps
+    info('validating')
+    for p in steps:
+        p.validate()
+
+    successful_steps = list()
+    for step in steps:
```

Comment:
firstly, mv this into the transaction class (see above).
secondly, you also must handle any exceptions occurring during reversion. we must try to revert as much as possible, even if errors occur in the process